File: cpp/LinkedList/singly/list_test.cc
#include <iostream>
#include <assert.h>
#include "list.h"  // Adjust the include path as needed

void testPrepend() {
    LinkedList<int> list;
    list.prepend(1);
    assert(list.getHead()->data == 1);
    list.prepend(0);
    assert(list.getHead()->data == 0);
    assert(list.getHead()->next->data == 1);
    std::cout << "testPrepend passed!" << std::endl;
}

void testAppend() {
    LinkedList<int> list;
    list.append(1);
    assert(list.getTail()->data == 1);
    list.append(2);
    assert(list.getTail()->data == 2);
    std::cout << "testAppend passed!" << std::endl;
}

void testInsertBetween() {
    LinkedList<int> list;
    list.append(1);
    list.append(3);
    list.insertBetween(2, 3, 1);
    assert(list.getHead()->next->data == 2);
    assert(list.getHead()->next->next->data == 3);
    std::cout << "testInsertBetween passed!" << std::endl;
}

void testDeleteHead() {
    LinkedList<int> list;
    list.append(1);
    list.append(2);
    list.deleteHead();
    assert(list.getHead()->data == 2);
    std::cout << "testDeleteHead passed!" << std::endl;
}

void testDeleteTail() {
    LinkedList<int> list;
    list.append(1);
    list.append(2);
    list.deleteTail();
    assert(list.getTail()->data == 1);
    std::cout << "testDeleteTail passed!" << std::endl;
}

void testDeleteAny() {
    LinkedList<int> list;
    list.append(1);
    list.append(2);
    list.append(3);
    list.deleteAny(2);
    assert(list.getHead()->next->data == 3);
    std::cout << "testDeleteAny passed!" << std::endl;
}

void testSize() {
    LinkedList<int> list;
    assert(list.size() == 0);
    list.append(1);
    list.append(2);
    list.append(3);
    assert(list.size() == 3);
    std::cout << "testSize passed!" << std::endl;
}

void testReverseList() {
    LinkedList<int> list;
    list.append(1);
    list.append(2);
    list.append(3);
    list.reverseList();
    assert(list.getHead()->data == 3);
    assert(list.getTail()->data == 1);
    std::cout << "testReverseList passed!" << std::endl;
}

void testFind() {
    LinkedList<int> list;
    list.append(1);
    list.append(2);
    assert(list.find(2)->data == 2);
    assert(list.find(3) == nullptr);
    std::cout << "testFind passed!" << std::endl;
}

void testClear() {
    LinkedList<int> list;
    list.append(1);
    list.append(2);
    list.clear();
    assert(list.getHead() == nullptr);
    assert(list.size() == 0);
    std::cout << "testClear passed!" << std::endl;
}

int main() {
    testPrepend();
    testAppend();
    testInsertBetween();
    testDeleteHead();
    testDeleteTail();
    testDeleteAny();
    testSize();
    testReverseList();
    testFind();
    testClear();

    std::cout << "All tests passed!" << std::endl;
    return 0;
}

File: python/linkedlist/singly/list_test.py
import unittest
from list import LinkedList


class TestLinkedList(unittest.TestCase):

    class TestData:
        def __init__(self, id, value):
            self.id = id
            self.value = value

        def __str__(self):
            return f"({self.id}: {self.value})"

    def setUp(self):
        self.linked_list = LinkedList()

    def test_prepend(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        self.linked_list.prepend(data1)
        self.linked_list.prepend(data2)
        self.assertEqual(self.linked_list.get_head().data, data2)
        self.assertEqual(self.linked_list.get_tail().data, data1)
        self.assertEqual(self.linked_list.size(), 2)

    def test_append(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        self.linked_list.append(data1)
        self.linked_list.append(data2)
        self.assertEqual(self.linked_list.get_head().data, data1)
        self.assertEqual(self.linked_list.get_tail().data, data2)
        self.assertEqual(self.linked_list.size(), 2)

    def test_insert_between(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        data3 = self.TestData(3, "data3")
        self.linked_list.append(data1)
        self.linked_list.append(data3)
        self.linked_list.insert_between(data2, before=data3, after=data1)
        self.assertEqual(self.linked_list.size(), 3)
        self.assertEqual(self.linked_list.get_head().next.data, data2)
        self.assertEqual(self.linked_list.get_head().next.next.data, data3)

    def test_delete_head(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        self.linked_list.append(data1)
        self.linked_list.append(data2)
        self.linked_list.delete_head()
        self.assertEqual(self.linked_list.get_head().data, data2)
        self.assertEqual(self.linked_list.size(), 1)
        self.linked_list.delete_head()
        self.assertIsNone(self.linked_list.get_head())
        self.assertIsNone(self.linked_list.get_tail())
        self.assertEqual(self.linked_list.size(), 0)

    def test_delete_tail(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        self.linked_list.append(data1)
        self.linked_list.append(data2)
        self.linked_list.delete_tail()
        self.assertEqual(self.linked_list.get_tail().data, data1)
        self.assertEqual(self.linked_list.size(), 1)
        self.linked_list.delete_tail()
        self.assertIsNone(self.linked_list.get_tail())
        self.assertIsNone(self.linked_list.get_head())
        self.assertEqual(self.linked_list.size(), 0)

    def test_delete_any(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        data3 = self.TestData(3, "data3")
        self.linked_list.append(data1)
        self.linked_list.append(data2)
        self.linked_list.append(data3)
        self.linked_list.delete_any(data2)
        self.assertEqual(self.linked_list.size(), 2)
        self.assertEqual(self.linked_list.get_head().next.data, data3)
        self.linked_list.delete_any(data1)
        self.assertEqual(self.linked_list.size(), 1)
        self.assertEqual(self.linked_list.get_head().data, data3)
        self.linked_list.delete_any(data3)
        self.assertEqual(self.linked_list.size(), 0)
        self.assertIsNone(self.linked_list.get_head())
        self.assertIsNone(self.linked_list.get_tail())

    def test_reverse_list(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        data3 = self.TestData(3, "data3")
        self.linked_list.append(data1)
        self.linked_list.append(data2)
        self.linked_list.append(data3)
        self.linked_list.reverse_list()
        self.assertEqual(self.linked_list.get_head().data, data3)
        self.assertEqual(self.linked_list.get_head().next.data, data2)
        self.assertEqual(self.linked_list.get_head().next.next.data, data1)
        self.assertEqual(self.linked_list.get_tail().data, data1)

    def test_find(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        self.linked_list.append(data1)
        self.linked_list.append(data2)
        self.assertEqual(self.linked_list.find(1), self.linked_list.get_head())
        self.assertEqual(self.linked_list.find(2), self.linked_list.get_tail())
        self.assertIsNone(self.linked_list.find(3))

    def test_clear(self):
        data1 = self.TestData(1, "data1")
        data2 = self.TestData(2, "data2")
        self.linked_list.append(data1)
        self.linked_list.append(data2)
        self.linked_list.clear()
        self.assertIsNone(self.linked_list.get_head())
        self.assertIsNone(self.linked_list.get_tail())
        self.assertEqual(self.linked_list.size(), 0)
        self.assertEqual(self.linked_list.node_map, {})


if __name__ == "__main__":
    unittest.main()
